src/TraceEvent/TraceLog.cs

```diff
@@ -153,10 +153,227 @@ public static TraceLog OpenOrConvert(string etlOrEtlxFilePath, TraceLogOptions o
+        /// };
+        /// var session = client.StartEventPipeSession(providers, requestRundown: false);
+        /// var eventSource = TraceLog.CreateFromEventPipeSession(session, rundownSession);
+        /// eventSource.Process();
+        /// </example>
+        public static TraceLogEventSource CreateFromEventPipeSession(EventPipeSession session, EventPipeSession rundownSession = null)
+        {
+            var tlEventSource = CreateFromEventPipeEventSource(new EventPipeEventSource(session.EventStream));
+
+            if (rundownSession != null)
+            {
+                tlEventSource.TraceLog.ProcessInitialRundown(rundownSession);
+            }
+
+            return tlEventSource;
+        }
+
+        /// <summary>
+        /// From a EventPipeEventSource, create a real time TraceLog Event Source. Like an EventPipeEventSource a TraceLogEventSource
+        /// will deliver events in real time. However an TraceLogEventSource has an underlying Tracelog (which you can access with
+        /// the .Log Property) which lets you get at aggregated information (Processes, threads, images loaded, and perhaps most
+        /// importantly TraceEvent.CallStack() will work. Thus you can get real time stacks from events).
+        /// </summary>
+        /// <param name="rundownSource">
+        /// If given, the rundownSource is used to initialize module and method information.
+        /// This only makes sense in realtime sessions when you need to resolve function names.
+        /// </param>
+        public static TraceLogEventSource CreateFromEventPipeEventSource(EventPipeEventSource source, EventPipeSession rundownSource = null)
+        {
+            var traceLog = new TraceLog(source);
+            traceLog.rawEventSourceToConvert.AllEvents += traceLog.onAllEventPipeEventsRealTime;
+
+            if (rundownSource != null)
+            {
+                traceLog.ProcessInitialRundown(rundownSource);
+            }
+
             return traceLog.realTimeSource;
         }
 
+        private void ProcessInitialRundown(EventPipeSession session)
+        {
+            using (var source = new EventPipeEventSource(session.EventStream))
+            {
+                SetupInitialRundownCallbacks(source);
+
+                // Only stopping the session will cause the rundown events to be sent.
+                // However, we cannot stop it before starting to process the source.
+                // Therefore, we need attach to the first event and only then stop the session.
+                var completionSource = new TaskCompletionSource<bool>();
+                source.AllEvents += delegate (TraceEvent _)
+                {
+                    completionSource.TrySetResult(true);
+                };
+                var task = Task.Run(source.Process);
+                completionSource.Task.Wait();
+                session.Stop();
```

Comment:
@davmason, will this work to ensure that the entire end rundown is received?

Reply:
Yeah, just stopping the session is enough